File: backend/app/schemas/material_cards.py
from datetime import datetime

from pydantic import BaseModel, ConfigDict

from app.models.events import EventType
from app.models.users import Area
from app.schemas.dictionaries import MaterialSkuOut
from app.schemas.units import MaterialUnitOut


class MaterialEventOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    event_id: int
    unit_id: int
    event_type: EventType
    area: Area | None
    timestamp: datetime
    user_id: int
    width_mm: float
    from_length: float | None
    to_length: float | None
    from_cell: str | None
    to_cell: str | None
    quantity_delta_m: float


class MaterialCardOut(BaseModel):
    sku: MaterialSkuOut
    total_area_m2: float
    units: list[MaterialUnitOut]
    events: list[MaterialEventOut]

File: frontend/src/api/materialCards.ts
import { apiClient } from "./client";
import type { MaterialSku, MaterialUnit } from "./units";

export interface MaterialEvent {
  event_id: number;
  unit_id: number;
  event_type: string;
  area: string | null;
  timestamp: string;
  user_id: number;
  width_mm: number;
  from_length: number | null;
  to_length: number | null;
  from_cell: string | null;
  to_cell: string | null;
  quantity_delta_m: number;
}

export interface MaterialCard {
  sku: MaterialSku;
  total_area_m2: number;
  units: MaterialUnit[];
  events: MaterialEvent[];
}

export async function getMaterialCard(skuId: number): Promise<MaterialCard> {
  const { data } = await apiClient.get<MaterialCard>(`/material-cards/${skuId}`);
  return data;
}

// Раздел про производителя/толщину внутри карточки материала (не
// отдельным измерением) — карточка теперь по группе материал+цвет, skus —
// все производители и толщины этой группы сразу (каждый со своими
// толщиной/код-у-поставщика/родная-ширина/фото/аналоги).
export interface MaterialCardGroup {
  material: string;
  color: string;
  skus: MaterialSku[];
  total_area_m2: number;
  units: MaterialUnit[];
  events: MaterialEvent[];
}

export async function getMaterialCardByGroup(material: string, color: string): Promise<MaterialCardGroup> {
  const { data } = await apiClient.get<MaterialCardGroup>("/material-cards/by-group", { params: { material, color } });
  return data;
}
